python/cudf/cudf/utils/utils.py

```diff
@@ -473,6 +473,14 @@ def search_range(start, stop, x, step=1, side="left"):
     return max(min(length, i), 0)
 
 
+def is_cudf_pyscalar(obj):
```

Comment:
I'd much rather we didn't punt to utilities for small tasks like this:

```python
isinstance(obj, cudf.Scalar)  # very easy to reason about
is_cudf_pyscalar(obj)  # need to see implementation/docstring of is_cudf_scalar to understand
```

I understand that we're trying to avoid having to import things for performance reasons. At the same time, we also have issues relating to circular imports. I'd suggest one of the following approaches:

1. Don't worry about it: unless the dict lookup time really shows up in the profiles as a significant bottleneck. That has been the case in the past in specific parts of the codebase that are very frequently called.

2. Always pass a `cudf.Scalar` directly to the Cython layer, thus preventing the need to do an `isinstance()` check there. We can do the `isinstance` check in the caller, where there may not be any circular import problems.

Reply:
Moved the branching logic to python here